workspace_host/interface.js

```diff
@@ -235,11 +258,193 @@ watcher.on('unlinkDir', filename => {
+ * serially instead of in parallel.
+ */
+async function pushAllRunningContainersToS3() {
+    const result = await sqldb.queryAsync(sql.get_running_workspaces, { instance_id: workspace_server_settings.instance_id });
+    await async.eachSeries(result.rows, async (ws) => {
+        if (ws.state == 'running') {
+            await pushContainerContentsToS3(ws);
+        }
+    });
+}
+setInterval(pushAllRunningContainersToS3, config.workspaceHostForceUploadIntervalSec * 1000);
+
+/* Prune stopped and runaway containers */
+
+/**
+ * Remove any recently stopped containers.  These are identified by having a non-null launch_uuid
+ * and launch_port, but are marked as stopped by the main PrairieLearn instance.
+ */
+async function pruneStoppedContainers() {
+    const instance_id = workspace_server_settings.instance_id;
+    const recently_stopped = await sqldb.queryAsync(sql.get_stopped_workspaces, { instance_id });
+    await async.each(recently_stopped.rows, async (ws) => {
+        let container;
+        try {
+            /* Try to grab the container, but don't care if it doesn't exist */
+            container = await _getDockerContainerByLaunchUuid(ws.launch_uuid);
+        } catch (_err) {
+            /* No container */
+            await sqldb.queryAsync(sql.clear_workspace_on_shutdown, { workspace_id: ws.id, instance_id: workspace_server_settings.instance_id });
+            return;
+        }
+        await pushContainerContentsToS3(ws);
+        await dockerAttemptKillAndRemove(container);
+        await sqldb.queryAsync(sql.clear_workspace_on_shutdown, { workspace_id: ws.id, instance_id: workspace_server_settings.instance_id });
+    });
+}
 
-async function _getAvailablePort(workspace_id, lowest_usable_port) {
-    function _checkPortAvailability(port) {
+/**
+ * Remove any containers that are running but do not have a corresponding launch_uuid in the database.
+ */
+async function pruneRunawayContainers() {
+    const instance_id = workspace_server_settings.instance_id;
+    const db_workspaces = await sqldb.queryAsync(sql.get_running_workspaces, { instance_id });
+    const db_workspaces_uuid_set = new Set(db_workspaces.rows.map(ws => `workspace-${ws.launch_uuid}`));
+    let running_workspaces;
+    try {
+        running_workspaces = await docker.listContainers();
+    } catch (err) {
+        /* Nothing to do */
+        return;
+    }
+
+    await async.each(running_workspaces, async (container_info) => {
+        if (container_info.Names.length != 1) return;
+        const name = container_info.Names[0].substring(1); /* Remove the preceding forward slash */
+        if (!name.startsWith('workspace-') || db_workspaces_uuid_set.has(name)) return;
+        await dockerAttemptKillAndRemove(container_info.Id);
+    });
+}
+
+setInterval(async () => {
+    await pruneStoppedContainers();
+    await pruneRunawayContainers();
+}, config.workspaceHostPruneContainersSec * 1000);
```

Comment:
Technically we could have multiple copies of this running simultaneously?

Reply:
use a lock? :grinning: